runtime-sdk/src/modules/consensus/mod.rs

```diff
@@ -0,0 +1,204 @@
+//! Consensus module.
+use thiserror::Error;
+
+use oasis_core_runtime::consensus::{
+    roothash::{Message, StakingMessage},
+    staking,
+    staking::Account as ConsensusAccount,
+    state::staking::ImmutableState as StakingImmutableState,
+};
+
+use crate::{
+    context::Context,
+    types::{address::Address, message::MessageEventHookInvocation, token},
+};
+
+#[cfg(test)]
+mod test;
+
+const MODULE_NAME: &str = "consensus";
+
+#[derive(Error, Debug, oasis_runtime_sdk_macros::Error)]
+pub enum Error {
+    #[error("invalid argument")]
+    #[sdk_error(code = 1)]
+    InvalidArgument,
+}
+/// Interface that can be called from other modules.
+pub trait API {
+    /// Transfer an amount from the runtime account.
+    fn transfer<C: Context>(
+        ctx: &mut C,
+        to: Address,
+        amount: &token::BaseUnits,
+        hook: MessageEventHookInvocation,
+    ) -> Result<(), Error>;
+
+    /// Withdraw an amount into the runtime account.
+    fn withdraw<C: Context>(
+        ctx: &mut C,
+        from: Address,
+        amount: &token::BaseUnits,
+        hook: MessageEventHookInvocation,
+    ) -> Result<(), Error>;
+
+    /// Escrow an amount of the runtime account.
+    fn escrow<C: Context>(
+        ctx: &mut C,
+        to: Address,
+        amount: &token::BaseUnits,
+        hook: MessageEventHookInvocation,
+    ) -> Result<(), Error>;
+
+    /// Reclaim an amount of runtime staked shares from the account.
+    fn reclaim_escrow<C: Context>(
+        ctx: &mut C,
+        from: Address,
+        amount: &token::BaseUnits,
+        hook: MessageEventHookInvocation,
+    ) -> Result<(), Error>;
+
+    // TODO:
+    //  - more staking queries.
+
+    /// Query consensus account info.
+    fn account<C: Context>(ctx: &C, addr: Address) -> Result<ConsensusAccount, Error>;
+}
+
+/// Consensus module.
+///
+/// Low level consensus module supporting runtime interactions with the consensus layer.
+pub struct Module {}
+
+impl Module {
+    fn do_transfer<C: Context>(
+        ctx: &mut C,
+        addr: Address,
+        amount: &token::BaseUnits,
+        hook: MessageEventHookInvocation,
+    ) -> Result<(), Error> {
+        // TODO: should probably enforce denomination here. Ensure it matches consensus token symbol?
+        ctx.emit_message(
+            Message::Staking {
+                v: 0,
+                msg: StakingMessage::Transfer(staking::Transfer {
+                    to: addr.into(),
+                    amount: amount.amount().clone(),
+                }),
+            },
+            hook,
+        )
+        .map_err(|_| Error::InvalidArgument)
```

Comment:
I think there should be a more specific error (defined in the core module) for "out of message slots" (similar to out of gas). Later with https://github.com/oasisprotocol/oasis-core/issues/3760 we can have the transaction scheduler make sure that transactions are scheduled such that they don't go over this limit.

Reply:
Agreed. So emit_message already returned a similar error. I moved that error (and renamed it) to the core module. Also updated this consensus helper module to return module::core errors, so that the error can be used directly

Reply:
hm actually i still need a way to propagate this error in other modules (e.g. consensus_accounts module cannot return it currently cause it implements it's own error type...). Hm i guess module defined error enums should actually "extend" the core module errors?